lib/puppet/configurer.rb

```diff
@@ -168,6 +189,23 @@ def run_internal(options)
 
     Puppet::Util::Log.newdestination(report)
 
+    # If a cached catalog is explicitly requested, attempt to retrieve it. Skip the node request,
+    # don't pluginsync and switch to the catalog's environment if we successfully retrieve it.
+    if Puppet[:use_cached_catalog]
+      if catalog = prepare_and_retrieve_catalog_from_cache
+        options[:catalog] = catalog
+        @cached_catalog_status = 'explicitly_requested'
+        @environment = catalog.environment
+        report.environment = catalog.environment
+      else
+        # Don't try to retrieve a catalog from the cache again after we've already
+        # failed to do so the first time.
+        Puppet[:use_cached_catalog] = false
+        Puppet[:usecacheonfailure] = false
+        options[:pluginsync] = Puppet::Configurer.should_pluginsync?
```

Comment:
This makes sense, no need to try the cache again, unless some code below has a side-effect of storing a cached catalog. But I don't think that's the case.